### ISSUE TITLE
Multicast documentation is misleading

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4262]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#4262|
|Type|defect|
|Created|2010-02-08 17:16:46Z|
|Branch|https://github.com/twisted/twisted/tree/multicast-doc-4262|

1. The client example in the multicast docs doesn't really explain that it will not be able to receive multicast messages, only send them. The second example should therefore probably also have a joinGroup() in the example.

2. The server example shows replying via UDP, rather than multicast. There should be some explanation about how it's possible to do both.

3. The comment in the server example about doing .join() may be wrong, it should probably also be a joinGroup().

Attachments:

* [patch-multicastdoc-4262.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/275/275dc65644b29251f4a77d6bb801d7f63f4ffec1/ec39dfb5cad7472052460aa1a12df5c652a0443a.patch) (8397 bytes) - added by frap on 2011-11-07 20:05:33Z - 
* [patch-mcastdoc-4262-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/275/275dc65644b29251f4a77d6bb801d7f63f4ffec1/e7db2118b7efb4edddcbde220115236335abf099.patch) (1690 bytes) - added by frap on 2011-11-10 11:02:50Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4262 4262
type__defect defect
reporter__itamar itamar
priority__normal normal
milestone__ 
branch__branches_multicast_doc_4262 branches/multicast-doc-4262
branch_author__itamarst__frap itamarst, frap
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation__easy documentation, easy
time__1265649406000000 1265649406000000
changetime__1321215715000000 1321215715000000
version__None None
owner__itamar itamar
cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph set owner to @itamarst|
|-|-|

Unfortunately I don't really know anything about multicast; you'll have to find someone who does to fix the docs.

**Comment by twisted-trac** [comment]
|<img alt="frap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|frap set owner to frap<br>frap set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="frap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|frap commented|
|-|-|

1: Updated - second example was using UDP unicast for reply so have changed to multicast
2. Not sure what you mean here via Ethernet (L2) multicast? Multicast for IP needs a transport and that is UDP?
3. have updated as well as changed to xmtml documentation to better reflect how multicast works

I aslo noticed there are no tests for multicast transport and its special methods like JoinGroup, setTTL - should I look at trying to do that? Would I have to mock these?

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Thanks, putting it up for review (by assigning it to nobody, see ReviewProcess#Authors:Howtogetyourchangereviewed). Please ask on the #twisted irc channel about the missing tests and/or open a new ticket for that.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

There are actually tests for multicast in `twisted.test.test_udp`.

**Comment by twisted-trac** [comment]
|<img alt="frap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|frap commented|
|-|-|

Replying to [itamar](#note_5):
> There are actually tests for multicast in `twisted.test.test_udp`.

yeah my bad didnt realise test for core functionality were above core directory. Hopefully rest of changes are suitable

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

(In [33099]) Branching to 'multicast-doc-4262'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_8"></a>|@itamarst commented|
|-|-|

I've simplified and edited the patch enough that I probably shouldn't be reviewer, so leaving with review keyword.

**Comment by twisted-trac** [comment]
|<img alt="frap's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|frap set owner to @itamarst|
|-|-|

Changes look good, I have just edited the comments of the code samples to make them clearer as the client wasn't unicasting back. Also changed the clients datagramReceived method so it prints the address so people runnig the code can see they receive their own ping.

If you are happy with the changes then please merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst removed owner|
|-|-|

Patch applied. I'll see if I can get anyone else to glance at this.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_12"></a>|@jesstess set owner to @itamarst|
|-|-|

This looks great! Thanks for working on this frap and itamar. A few tiny things:

 * "This in contrast with normal" ==> "This is in contrast with normal"
 * "(the IPv4 range 224.0.0.0 to 239.255.255.255)" ==> "(in the IPv4 range 224.0.0.0 to 239.255.255.255)", or "(the IPv4 range is 224.0.0.0 to 239.255.255.255)" ?

Other than that, I think this is ready to merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_13"></a>|@itamarst set status to `closed`|
|-|-|

(In [33121]) Merge multicast-doc-4262.

Author: frap, itamar
Review: jesstess, itamar
Fixes: [#4262](https://github.com/twisted/twisted/issues/4262)

Improve the multicast documentation, with better explanations and sample code.
